[JSC] JSBigInt division / remainder should have a fast path for same size

Constellation opened the pull request:
#### 40830da5de3d57fed8dcff3c0314b9b7986e2420
<pre>
[JSC] JSBigInt division / remainder should have a fast path for same size
<a href="https://bugs.webkit.org/show_bug.cgi?id=304380">https://bugs.webkit.org/show_bug.cgi?id=304380</a>
<a href="https://rdar.apple.com/166757372">rdar://166757372</a>

Reviewed by Keith Miller.

When there is `a / b`, if `a &lt; b`, then the result is 0.
If `a == b`, the result is 1. And if `a &gt; b` but `a` has the same number
of digits as `b`, then the quotient/result is one Digit.
By leveraging this, we can largely simplify the division algorithm.
This patch adds divideSameSize and remainderSameSize. And they completely
avoid any Vector allocation for temporary storage and compute the result
in an in-place manner.

Because we are pruning a path for `a &lt; b` and `a == b` before, same size
division is very common. So this is worth optimizing.

We also clean up a bit of span functions to remove assertions. We use
size_t and RELEASE_ASSERT to confirm these assertions are always true.

* Source/JavaScriptCore/runtime/JSBigInt.cpp:
(JSC::normalize):
(JSC::JSBigInt::multiplySingle):
(JSC::JSBigInt::multiplyTextbook):
(JSC::JSBigInt::divideSingle):
(JSC::JSBigInt::addAndReturnCarry):
(JSC::JSBigInt::subtractAndReturnBorrow):
(JSC::JSBigInt::leftShift):
(JSC::JSBigInt::rightShift):
(JSC::JSBigInt::divideTextbook):
(JSC::estimateQhat):
(JSC::JSBigInt::divideSameSize):
(JSC::JSBigInt::remainderSameSize):
(JSC::JSBigInt::divideImpl):
(JSC::JSBigInt::remainderImpl):
(JSC::JSBigInt::multiplyAdd):
(JSC::JSBigInt::addTextbook):
(JSC::JSBigInt::subTextbook):
(JSC::JSBigInt::absoluteBitwiseOp):
(JSC::JSBigInt::absoluteAndNot):
(JSC::JSBigInt::absoluteAddOne):
(JSC::JSBigInt::absoluteSubOne):
(JSC::JSBigInt::leftShiftByAbsolute):
(JSC::JSBigInt::rightShiftByAbsolute):
* Source/JavaScriptCore/runtime/JSBigInt.h:

Canonical link: <a href="https://commits.webkit.org/304712@main">https://commits.webkit.org/304712@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/219e8533c38eb00d75ccbd189a1a61e416caffe8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136366 "6 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8723 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47646 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144078 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ea01d0d9-8d24-4ae8-89aa-51851a75e381) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9405 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8567 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/104273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/948372ff-26cc-488a-b2cc-773432d1f318) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139311 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/6851 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/122188 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/85109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/6501 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/4160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4669 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/128323 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/115795 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/40390 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146822 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/134850 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8405 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/40958 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/112613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8422 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/7058 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/112958 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/6429 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/118493 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21017 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8453 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/36548 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/167629 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8171 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/72012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/43728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8393 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8245 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/runtime/JSBigInt.cpp:
```diff
@@ -1255,6 +1259,148 @@ std::tuple<std::span<JSBigInt::Digit>, std::span<JSBigInt::Digit>> JSBigInt::div
+    // a.back() > b.back(), so a > b and quotient is at least 1.
+    // Since a and b have the same number of digits, the quotient fits in one digit.
+    //
+    // We estimate qhat by normalizing only the top 2-3 digits (no vector allocation),
```
Probably worth saying "qhat from Algorithm D"

Constellation replied:
Added.

kmiller68 commented on Source/JavaScriptCore/runtime/JSBigInt.cpp:
```diff
@@ -1255,6 +1259,148 @@ std::tuple<std::span<JSBigInt::Digit>, std::span<JSBigInt::Digit>> JSBigInt::div
     return { qResult, rResult };
 }
 
+static ALWAYS_INLINE JSBigInt::Digit estimateQhat(std::span<const JSBigInt::Digit> a, std::span<const JSBigInt::Digit> b)
```
It's kinda a shame that we use a/b everywhere rather than numerator and divisor or some equivalents.

kmiller68 reviewed (approved):
r=me, this could use some tests of interesting edge cases e.g. `(1 << 63) / 2` though. 

I also think the commit message is wrong:
```And if `a > b`, then the quotient is one Digit. So result is always one digit, and by leveraging this, we can largely simplify the division algorithm.``` 
should be 
```And if `a > b` but `a` has the same number of digits as `b`, then the quotient/result is one Digit. By leveraging this, we can largely simplify the division algorithm.``` I was very confused until I started reading the code.

Constellation committed:
[JSC] JSBigInt division / remainder should have a fast path for same size
https://bugs.webkit.org/show_bug.cgi?id=304380
rdar://166757372

Reviewed by Keith Miller.

When there is `a / b`, if `a < b`, then the result is 0.
If `a == b`, the result is 1. And if `a > b` but `a` has the same number
of digits as `b`, then the quotient/result is one Digit.
By leveraging this, we can largely simplify the division algorithm.
This patch adds divideSameSize and remainderSameSize. And they completely
avoid any Vector allocation for temporary storage and compute the result
in an in-place manner.

Because we are pruning a path for `a < b` and `a == b` before, same size
division is very common. So this is worth optimizing.

We also clean up a bit of span functions to remove assertions. We use
size_t and RELEASE_ASSERT to confirm these assertions are always true.

* Source/JavaScriptCore/runtime/JSBigInt.cpp:
(JSC::normalize):
(JSC::JSBigInt::multiplySingle):
(JSC::JSBigInt::multiplyTextbook):
(JSC::JSBigInt::divideSingle):
(JSC::JSBigInt::addAndReturnCarry):
(JSC::JSBigInt::subtractAndReturnBorrow):
(JSC::JSBigInt::leftShift):
(JSC::JSBigInt::rightShift):
(JSC::JSBigInt::divideTextbook):
(JSC::estimateQhat):
(JSC::JSBigInt::divideSameSize):
(JSC::JSBigInt::remainderSameSize):
(JSC::JSBigInt::divideImpl):
(JSC::JSBigInt::remainderImpl):
(JSC::JSBigInt::multiplyAdd):
(JSC::JSBigInt::addTextbook):
(JSC::JSBigInt::subTextbook):
(JSC::JSBigInt::absoluteBitwiseOp):
(JSC::JSBigInt::absoluteAndNot):
(JSC::JSBigInt::absoluteAddOne):
(JSC::JSBigInt::absoluteSubOne):
(JSC::JSBigInt::leftShiftByAbsolute):
(JSC::JSBigInt::rightShiftByAbsolute):
* Source/JavaScriptCore/runtime/JSBigInt.h:

Canonical link: https://commits.webkit.org/304712@main
